### PR TITLE
[MediaBundle] Azure Blob Storage implementation

### PR DESCRIPTION
**Pull request by plozmun**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | -
| Related issues/PRs |-
| License | MIT
| Documentation PR | sulu/sulu-docs#502

#### What's in this PR?

 Azure Blob Storage implementation

#### Why?

We have a proposal from a potential client that requires the use of azure. More and more companies are using the Microsoft cloud

#### Example Usage

See docs

#### To Do

- [X] Create a documentation PR
- [ ] Add breaking changes to UPGRADE.md

I suggest to refactor s3 Storage because in 2.0 access to client in s3 will be removed ( see https://github.com/thephpleague/flysystem-azure-blob-storage/pull/18 )

### CODE REVIEW

**Comment by plozmun** [comment]
AppVeyor fails for execution time and travis doesn't seem to want to start 😅

File: src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -224,6 +227,21 @@ private function addStorageSection(ArrayNodeDefinition $node)
                 ->end();
         }
 
+        if (class_exists(AzureBlobStorageAdapter::class)) {
+            $storages[] = self::STORAGE_AZURE_BLOB;
+
+            $storagesNode
+                ->arrayNode(self::STORAGE_AZURE_BLOB)
+                ->addDefaultsIfNotSet()
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php]: Can you indent this lines to match the tree like in the other configuraitons

Comment (plozmun) [reply src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php]: Done!

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php]: This line should also be indented. So L235-L241 should be +4 spaces. See S3 example.

Comment (plozmun) [reply src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php]: I didn't realize 😅 , I think it's already

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/DependencyInjection/Configuration.php]: Thank you!

File: src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        $this->container = $container;
+        $this->adapter = $filesystem->getAdapter();
+    }
+
+    public function load(array $storageOptions)
+    {
+        $resource = parent::load($storageOptions);
+        $contents = stream_get_contents($resource);
+
+        $stream = fopen('php://memory', 'r+');
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php]: @plozmun @alexander-schranz why do have to copy it to the memory?

**Comment by wachterjohannes** [comment]
@alexander-schranz @plozmun i have added a question to the code - anything else looks good

but i also have no azure account

File: src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php (hunk 2 above)
Comment (plozmun) [reply src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php]: Azure Filesystem returns a not seekable object, not like Google Filesystem than returns a PSR7Message seekable, then this fails when it tries to generate thumbnail https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/ImagineImageConverter.php#L152
due to this https://github.com/sulu/sulu/blob/master/src/Sulu/Bundle/MediaBundle/Media/ImageConverter/MediaImageExtractor.php#L52 mime_content_type doest't support not seekable objects

Comment (wachterjohannes) [reply src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php]: OK sounds logic to me - i would at least add a comment for this to be sure that we do not remove it intentionally.

Comment (plozmun) [reply src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php]: Done! Check the comment and tell me if it's enough

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php]: Can we add the comment before `$contents = stream_get_contents($resource);` on the method its mostly overlooked.

Would also add more information to it e.g.:

```php
// Azure Filesystem returns a not seekable resource
// Need convert to a seekable resource to allow get mimetype from it for MediaImageExtractor
$contents = stream_get_contents($resource);
...
```

Comment (alexander-schranz) [reply src/Sulu/Bundle/MediaBundle/Media/Storage/AzureBlobStorage.php]: @plozmun thank you!

**Comment by plozmun** [comment]
Have you been able to test this? We have it in pre-production with a git patch and everything is correct.

**Comment by alexander-schranz** [comment]
Not sure why travis is not triggered for this pull request I tested it locally and the tests did work, so I decided to merge it manually.

@plozmun Thank you for this contribution 👍

**Pull request merged**